src/directives/run-async.ts

```diff
@@ -0,0 +1,153 @@
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {directive, NodePart, Part} from '../lit-html.js';
+
+interface AsyncRunState {
+  key: unknown;
+  promise: Promise<unknown>;
+  abortController?: AbortController;
+  state: 'initial'|'pending'|'success'|'failure';
+  resolvePending: () => void;
+  rejectPending: (e: Error) => void;
+}
+
+const hasAbortController = typeof AbortController === 'function';
+
+const runs = new WeakMap<Part, AsyncRunState>();
+
+export type Task<K> = (key: K, options: {signal?: AbortSignal}) =>
+    Promise<unknown>;
+
+/**
+ * Runs an async function whenever the key changes, and calls one of several
+ * lit-html template functions depending on the state of the async call:
+ *
+ *  - success() is called when the result of the function resolves.
+ *  - pending() is called immediately
+ *  - initial() is called if the function rejects with a InitialStateError,
+ *    which lets the function indicate that it couldn't proceed with the
+ *    provided key. This is usually the case when there isn't data to load.
+ *  - failure() is called if the function rejects.
+ *
+ * @param key A parameter passed to the task function. The task function is only
+ *     called when they key changes.
+ * @param task An async function to run when the key changes
+ * @param templates The templates to render for each state of the task
+ */
+export const runAsync = directive(<K>(key: K, task: Task<K>, templates: {
+                                    success: (result: any) => any,
+                                    pending?: () => any,
+                                    initial?: () => any,
+                                    failure?: (e: Error) => any
+                                  }) => (part: NodePart) => {
+  const {success, pending, initial, failure} = templates;
+  const currentRunState = runs.get(part);
+
+  // The first time we see a value we save and await the work function.
+  // TODO(justinfagnani): allow a custom invalidate function
+  if (currentRunState === undefined || currentRunState.key !== key) {
+    // Abort a pending request if there is one
+    if (currentRunState !== undefined && currentRunState.state === 'pending') {
+      if (currentRunState.abortController !== undefined) {
+        currentRunState.abortController.abort();
+      }
+      // TODO(justinfagnani): This should be an AbortError, but it's not
+      // implemented yet
+      currentRunState.rejectPending(new Error());
+    }
+    const abortController =
+        hasAbortController ? new AbortController() : undefined;
+    const abortSignal =
+        hasAbortController ? abortController!.signal : undefined;
+    let resolvePending!: () => void;
+    let rejectPending!: (e: Error) => void;
+    const pendingPromise = new Promise((res, rej) => {
+      resolvePending = res;
+      rejectPending = rej;
+    });
+    const promise = task(key, {signal: abortSignal});
+    // The state is immediately 'pending', since the function has been
```

Comment:
Does this mean that even if the `promise` can't be run with the data available it will take always take two renders to get to `initial`? One pass for `pending` and then a second (assuming `throw new InitialStateError()`) to trigger `initial`. If so, that seems confusing to end users.

This may be conflating the responsibilities of `lit-html` and a component system (i.e. `LitElement) that relies on it, but I would expect more mainstream usage of this to rely on a key that powered the promise in someway so an initial call where `key === undefined` or `key === ''` would provide the `initial` render template in a single render pass.

In this case, would you expect there to be additional wrapping of or an alternative to this directive to achieve that functionality? or, would it make sense to include that by default?